File: src/python/common/core/service_discovery/network.py
"""
Network-based Service Discovery

This module implements UDP multicast-based network discovery for services
on the local network. It allows services to announce themselves and discover
other services without relying on a central registry file.
"""

import asyncio
import json
import socket
import struct
import time
import uuid
from dataclasses import dataclass, asdict
from datetime import datetime, timezone
from enum import Enum
from typing import Dict, List, Optional, Union, Callable
from loguru import logger

from .exceptions import NetworkError
from .registry import ServiceInfo

# logger imported from loguru


class DiscoveryMessageType(Enum):
    """Discovery message types"""
    DISCOVERY_REQUEST = "discovery_request"
    DISCOVERY_RESPONSE = "discovery_response"
    HEALTH_PING = "health_ping"
    SERVICE_ANNOUNCEMENT = "service_announcement"
    SERVICE_SHUTDOWN = "service_shutdown"


@dataclass
class DiscoveryMessage:
    """Network discovery message"""
    message_type: DiscoveryMessageType
    service_name: str
    timestamp: str
    payload: Dict
    auth_token: Optional[str] = None

    def to_json(self) -> str:
        """Convert to JSON string"""
        data = asdict(self)
        data['message_type'] = self.message_type.value
        return json.dumps(data)

    @classmethod
    def from_json(cls, json_str: str) -> 'DiscoveryMessage':
        """Create from JSON string"""
        data = json.loads(json_str)
        data['message_type'] = DiscoveryMessageType(data['message_type'])
        return cls(**data)


@dataclass
class DiscoveryEvent:
    """Discovery events broadcast to subscribers"""
    event_type: str
    service_name: str
    service_info: Optional[ServiceInfo] = None
    status: Optional[str] = None
    metrics: Optional[Dict[str, str]] = None


class NetworkDiscovery:
    """Network discovery manager using UDP multicast"""

    def __init__(self, 
                 multicast_address: str = "239.255.42.42",
                 port: int = 9999,
                 auth_token: Optional[str] = None):
        """Initialize network discovery"""
        
        # Validate multicast address
        try:
            addr_bytes = socket.inet_aton(multicast_address)
            # Check if it's in multicast range (224.0.0.0 to 239.255.255.255)
            first_octet = addr_bytes[0]
            if not (224 <= first_octet <= 239):
                raise NetworkError(f"{multicast_address} is not a valid multicast address")
        except socket.error:
            raise NetworkError(f"Invalid multicast address: {multicast_address}")

        self.multicast_address = multicast_address
        self.port = port
        self.auth_token = auth_token
        self.socket: Optional[socket.socket] = None
        self.running = False
        self.discovered_services: Dict[str, tuple[ServiceInfo, float]] = {}
        self.event_callbacks: List[Callable[[DiscoveryEvent], None]] = []
        self.cache_timeout = 300.0  # 5 minutes
        
        logger.info(f"Network discovery initialized for {multicast_address}:{port}")

    async def start(self) -> None:
        """Start the network discovery service"""
        logger.info(f"Starting network discovery on {self.multicast_address}:{self.port}")

        try:
            # Create UDP socket
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            
            # Bind to multicast address
            self.socket.bind(('', self.port))
            
            # Join multicast group
            mreq = struct.pack("4sl", socket.inet_aton(self.multicast_address), socket.INADDR_ANY)
            self.socket.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)
            
            # Set socket to non-blocking
            self.socket.setblocking(False)
            
            self.running = True
            
            # Start receiver task
            asyncio.create_task(self._receive_loop())
            
            # Start cache cleanup task
            asyncio.create_task(self._cache_cleanup_loop())
            
            logger.info("Network discovery service started successfully")
            
        except Exception as e:
            raise NetworkError(f"Failed to start network discovery: {e}")

    async def stop(self) -> None:
        """Stop the network discovery service"""
        logger.info("Stopping network discovery service")
        
        self.running = False
        
        if self.socket:
            try:
                # Leave multicast group
                mreq = struct.pack("4sl", socket.inet_aton(self.multicast_address), socket.INADDR_ANY)
                self.socket.setsockopt(socket.IPPROTO_IP, socket.IP_DROP_MEMBERSHIP, mreq)
            except Exception as e:
                logger.warning(f"Error leaving multicast group: {e}")
            finally:
                self.socket.close()
                self.socket = None
        
        logger.info("Network discovery service stopped")

    async def announce_service(self, service_name: str, service_info: ServiceInfo) -> None:
        """Announce service availability"""
        logger.debug(f"Announcing service: {service_name}")
        
        message = DiscoveryMessage(
            message_type=DiscoveryMessageType.SERVICE_ANNOUNCEMENT,
            service_name=service_name,
            timestamp=datetime.now(timezone.utc).isoformat(),
            payload={"service_info": asdict(service_info)},
            auth_token=self.auth_token
        )
        
        await self._send_message(message)
        logger.debug(f"Service {service_name} announced successfully")

    async def discover_services(self, service_names: List[str], timeout: float = 10.0) -> Dict[str, ServiceInfo]:
        """Discover services on the network"""
        logger.info(f"Discovering services: {service_names}")
        
        request_id = str(uuid.uuid4())
        message = DiscoveryMessage(
            message_type=DiscoveryMessageType.DISCOVERY_REQUEST,
            service_name="discovery-client",
            timestamp=datetime.now(timezone.utc).isoformat(),
            payload={
                "service_names": service_names,
                "request_id": request_id
            },
            auth_token=self.auth_token
        )
        
        # Send discovery request
        await self._send_message(message)
        
        # Wait for responses
        start_time = time.time()
        discovered = {}
        
        while (time.time() - start_time) < timeout:
            # Check cached services
            for name, (service_info, last_seen) in list(self.discovered_services.items()):
                if (not service_names or name in service_names) and name not in discovered:
                    if time.time() - last_seen < self.cache_timeout:
                        discovered[name] = service_info
            
            # Break if we found all requested services
            if service_names and all(name in discovered for name in service_names):
                break
                
            await asyncio.sleep(0.1)
        
        logger.info(f"Discovery completed, found {len(discovered)} services")
        return discovered

    async def send_health_ping(self, service_name: str, status: str, metrics: Optional[Dict[str, str]] = None) -> None:
        """Send health ping"""
        logger.debug(f"Sending health ping for service: {service_name}")
        
        message = DiscoveryMessage(
            message_type=DiscoveryMessageType.HEALTH_PING,
            service_name=service_name,
            timestamp=datetime.now(timezone.utc).isoformat(),
            payload={
                "status": status,
                "metrics": metrics or {}
            },
            auth_token=self.auth_token
        )
        
        await self._send_message(message)

    async def announce_shutdown(self, service_name: str) -> None:
        """Announce service shutdown"""
        logger.info(f"Announcing shutdown for service: {service_name}")
        
        message = DiscoveryMessage(
            message_type=DiscoveryMessageType.SERVICE_SHUTDOWN,
            service_name=service_name,
            timestamp=datetime.now(timezone.utc).isoformat(),
            payload={},
            auth_token=self.auth_token
        )
        
        await self._send_message(message)

    def get_cached_services(self) -> Dict[str, ServiceInfo]:
        """Get currently discovered services from cache"""
        return {
            name: service_info 
            for name, (service_info, _) in self.discovered_services.items()
        }

    def subscribe_events(self, callback: Callable[[DiscoveryEvent], None]) -> None:
        """Subscribe to discovery events"""
        self.event_callbacks.append(callback)

    def unsubscribe_events(self, callback: Callable[[DiscoveryEvent], None]) -> None:
        """Unsubscribe from discovery events"""
        if callback in self.event_callbacks:
            self.event_callbacks.remove(callback)

    async def _send_message(self, message: DiscoveryMessage) -> None:
        """Send a message via multicast"""
        if not self.socket or not self.running:
            raise NetworkError("Socket not initialized")

        try:
            message_data = message.to_json().encode('utf-8')
            self.socket.sendto(message_data, (self.multicast_address, self.port))
            
            logger.debug(f"Sent {message.message_type.value} message for service {message.service_name}")
            
        except Exception as e:
            raise NetworkError(f"Failed to send message: {e}")

    async def _receive_loop(self) -> None:
        """Receive and process discovery messages"""
        while self.running:
            try:
                # Use select to check for available data
                ready = await asyncio.get_event_loop().run_in_executor(
                    None, self._wait_for_data, 0.1
                )
                
                if ready and self.socket:
                    data, addr = self.socket.recvfrom(4096)
                    try:
                        message = DiscoveryMessage.from_json(data.decode('utf-8'))
                        await self._process_message(message, addr)
                    except (json.JSONDecodeError, ValueError) as e:
                        logger.warning(f"Invalid message from {addr}: {e}")
                        
            except Exception as e:
                if self.running:  # Only log if we're supposed to be running
                    logger.warning(f"Error in receive loop: {e}")
                    
            await asyncio.sleep(0.001)  # Small delay to prevent busy waiting

    def _wait_for_data(self, timeout: float) -> bool:
        """Wait for socket data with timeout"""
        if not self.socket:
            return False
            
        import select
        ready, _, _ = select.select([self.socket], [], [], timeout)
        return bool(ready)

    async def _process_message(self, message: DiscoveryMessage, sender_addr: tuple) -> None:
        """Process received discovery message"""
        logger.debug(f"Processing {message.message_type.value} message from {sender_addr} for service {message.service_name}")

        # Verify authentication token if required
        if self.auth_token and message.auth_token != self.auth_token:
            logger.warning(f"Authentication failed for message from {sender_addr}")
            return

        try:
            if message.message_type == DiscoveryMessageType.SERVICE_ANNOUNCEMENT:
                service_info_data = message.payload.get("service_info", {})
                if service_info_data:
                    # Convert status back to enum if it's a string
                    if "status" in service_info_data and isinstance(service_info_data["status"], str):
                        from .registry import ServiceStatus
                        service_info_data["status"] = ServiceStatus(service_info_data["status"])
                    
                    service_info = ServiceInfo(**service_info_data)
                    self._cache_discovered_service(message.service_name, service_info)
                    
                    event = DiscoveryEvent(
                        event_type="service_discovered",
                        service_name=message.service_name,
                        service_info=service_info
                    )
                    self._broadcast_event(event)

            elif message.message_type == DiscoveryMessageType.DISCOVERY_REQUEST:
                # TODO: Respond with our own service information if applicable
                logger.debug(f"Received discovery request from {sender_addr}")

            elif message.message_type == DiscoveryMessageType.HEALTH_PING:
                status = message.payload.get("status", "unknown")
                metrics = message.payload.get("metrics", {})
                
                event = DiscoveryEvent(
                    event_type="health_ping",
                    service_name=message.service_name,
                    status=status,
                    metrics=metrics
                )
                self._broadcast_event(event)

            elif message.message_type == DiscoveryMessageType.SERVICE_SHUTDOWN:
                self._remove_cached_service(message.service_name)
                
                event = DiscoveryEvent(
                    event_type="service_lost",
                    service_name=message.service_name
                )
                self._broadcast_event(event)

        except Exception as e:
            logger.error(f"Error processing message: {e}")

    def _cache_discovered_service(self, service_name: str, service_info: ServiceInfo) -> None:
        """Cache a discovered service"""
        self.discovered_services[service_name] = (service_info, time.time())

    def _remove_cached_service(self, service_name: str) -> None:
        """Remove cached service"""
        self.discovered_services.pop(service_name, None)

    def _broadcast_event(self, event: DiscoveryEvent) -> None:
        """Broadcast event to all subscribers"""
        for callback in self.event_callbacks:
            try:
                callback(event)
            except Exception as e:
                logger.error(f"Error in event callback: {e}")

    async def _cache_cleanup_loop(self) -> None:
        """Cleanup expired cache entries"""
        while self.running:
            try:
                current_time = time.time()
                expired_services = []
                
                for service_name, (_, last_seen) in list(self.discovered_services.items()):
                    if current_time - last_seen > self.cache_timeout:
                        expired_services.append(service_name)
                
                for service_name in expired_services:
                    self._remove_cached_service(service_name)
                    event = DiscoveryEvent(
                        event_type="service_lost",
                        service_name=service_name
                    )
                    self._broadcast_event(event)
                
                if expired_services:
                    logger.debug(f"Cleaned up {len(expired_services)} expired service entries")
                    
            except Exception as e:
                logger.error(f"Error in cache cleanup: {e}")
                
            await asyncio.sleep(60)  # Clean every minute

File: src/rust/daemon/core/src/service_discovery/network.rs
//! Network-based Service Discovery
//!
//! This module implements UDP multicast-based network discovery for services
//! on the local network. It allows services to announce themselves and discover
//! other services without relying on a central registry file.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::net::{IpAddr, Ipv4Addr, SocketAddr, UdpSocket};
use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use thiserror::Error;
use tokio::sync::{broadcast, RwLock};
use tokio::time::{interval, timeout};
use tracing::{debug, error, info, warn};

use super::registry::ServiceInfo;

/// Network discovery errors
#[derive(Error, Debug)]
pub enum NetworkError {
    #[error("Socket bind error: {0}")]
    BindError(#[from] std::io::Error),
    
    #[error("Message serialization error: {0}")]
    SerializationError(#[from] serde_json::Error),
    
    #[error("Invalid multicast address: {0}")]
    InvalidMulticastAddress(String),
    
    #[error("Network timeout")]
    Timeout,
    
    #[error("Discovery request failed: {0}")]
    DiscoveryFailed(String),
}

/// Discovery message types
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum DiscoveryMessageType {
    /// Request to discover services
    DiscoveryRequest,
    /// Response with service information
    DiscoveryResponse,
    /// Health ping between services
    HealthPing,
    /// Service announcement
    ServiceAnnouncement,
    /// Service shutdown notification
    ServiceShutdown,
}

/// Network discovery message
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiscoveryMessage {
    /// Message type
    pub message_type: DiscoveryMessageType,
    
    /// Source service name
    pub service_name: String,
    
    /// Message timestamp (ISO 8601)
    pub timestamp: String,
    
    /// Message payload
    pub payload: DiscoveryPayload,
    
    /// Authentication token (optional)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auth_token: Option<String>,
}

/// Discovery message payload
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum DiscoveryPayload {
    /// Service information payload
    ServiceInfo(ServiceInfo),
    
    /// Request payload with filters
    Request {
        /// Requested service names (empty = all services)
        service_names: Vec<String>,
        /// Request ID for correlation
        request_id: String,
    },
    
    /// Health check payload
    Health {
        /// Service status
        status: String,
        /// Additional health metrics
        metrics: HashMap<String, String>,
    },
    
    /// Empty payload for simple messages
    Empty,
}

/// Network discovery manager
pub struct NetworkDiscovery {
    /// Multicast address for discovery
    multicast_addr: SocketAddr,
    
    /// UDP socket for sending/receiving messages
    socket: Arc<RwLock<Option<UdpSocket>>>,
    
    /// Discovered services cache
    discovered_services: Arc<RwLock<HashMap<String, (ServiceInfo, SystemTime)>>>,
    
    /// Message sender for broadcasting discovery events
    event_sender: broadcast::Sender<DiscoveryEvent>,
    
    /// Authentication token for secure communication
    auth_token: Option<String>,
    
    /// Cache timeout for discovered services
    cache_timeout: Duration,

    /// Locally registered services (services this node can respond for)
    local_services: Arc<RwLock<HashMap<String, ServiceInfo>>>,
}

/// Discovery events broadcast to subscribers
#[derive(Debug, Clone)]
pub enum DiscoveryEvent {
    /// Service discovered
    ServiceDiscovered {
        service_name: String,
        service_info: ServiceInfo,
    },
    
    /// Service updated
    ServiceUpdated {
        service_name: String,
        service_info: ServiceInfo,
    },
    
    /// Service lost (timeout or shutdown)
    ServiceLost {
        service_name: String,
    },
    
    /// Health ping received
    HealthPing {
        service_name: String,
        status: String,
        metrics: HashMap<String, String>,
    },
}

impl NetworkDiscovery {
    /// Create a new network discovery instance
    pub fn new(
        multicast_address: &str,
        port: u16,
        auth_token: Option<String>,
    ) -> Result<Self, NetworkError> {
        // Parse multicast address
        let multicast_ip: Ipv4Addr = multicast_address.parse()
            .map_err(|_| NetworkError::InvalidMulticastAddress(multicast_address.to_string()))?;
        
        // Validate multicast address range
        if !multicast_ip.is_multicast() {
            return Err(NetworkError::InvalidMulticastAddress(
                format!("{} is not a valid multicast address", multicast_address)
            ));
        }
        
        let multicast_addr = SocketAddr::new(IpAddr::V4(multicast_ip), port);
        let (event_sender, _) = broadcast::channel(1000);
        
        info!("Network discovery initialized for {}", multicast_addr);
        
        Ok(Self {
            multicast_addr,
            socket: Arc::new(RwLock::new(None)),
            discovered_services: Arc::new(RwLock::new(HashMap::new())),
            event_sender,
            auth_token,
            cache_timeout: Duration::from_secs(300), // 5 minutes
            local_services: Arc::new(RwLock::new(HashMap::new())),
        })
    }

    /// Start the network discovery service
    pub async fn start(&self) -> Result<(), NetworkError> {
        info!("Starting network discovery on {}", self.multicast_addr);
        
        // Create UDP socket
        let socket = UdpSocket::bind("0.0.0.0:0")?;
        socket.set_nonblocking(true)?;
        
        // Join multicast group
        let ip = self.multicast_addr.ip();
        if let IpAddr::V4(ipv4) = ip {
            socket.join_multicast_v4(&ipv4, &Ipv4Addr::UNSPECIFIED)?;
        } else {
            return Err(NetworkError::InvalidMulticastAddress("Multicast address must be IPv4".to_string()));
        }
        
        {
            let mut socket_guard = self.socket.write().await;
            *socket_guard = Some(socket);
        }
        
        // Start cache cleanup task
        self.start_cache_cleanup().await;
        
        info!("Network discovery service started successfully");
        Ok(())
    }

    /// Stop the network discovery service
    pub async fn stop(&self) -> Result<(), NetworkError> {
        info!("Stopping network discovery service");
        
        let mut socket_guard = self.socket.write().await;
        *socket_guard = None;
        
        info!("Network discovery service stopped");
        Ok(())
    }

    /// Announce service availability
    pub async fn announce_service(
        &self,
        service_name: &str,
        service_info: &ServiceInfo,
    ) -> Result<(), NetworkError> {
        debug!("Announcing service: {}", service_name);

        // Track as local service so we can respond to discovery requests
        {
            let mut local = self.local_services.write().await;
            local.insert(service_name.to_string(), service_info.clone());
        }

        let message = DiscoveryMessage {
            message_type: DiscoveryMessageType::ServiceAnnouncement,
            service_name: service_name.to_string(),
            timestamp: current_iso_timestamp(),
            payload: DiscoveryPayload::ServiceInfo(service_info.clone()),
            auth_token: self.auth_token.clone(),
        };

        self.send_message(&message).await?;

        debug!("Service {} announced successfully", service_name);
        Ok(())
    }

    /// Discover services on the network
    pub async fn discover_services(
        &self,
        service_names: Vec<String>,
        discover_timeout: Duration,
    ) -> Result<HashMap<String, ServiceInfo>, NetworkError> {
        info!("Discovering services: {:?}", service_names);
        
        let request_id = uuid::Uuid::new_v4().to_string();
        let message = DiscoveryMessage {
            message_type: DiscoveryMessageType::DiscoveryRequest,
            service_name: "discovery-client".to_string(),
            timestamp: current_iso_timestamp(),
            payload: DiscoveryPayload::Request {
                service_names: service_names.clone(),
                request_id: request_id.clone(),
            },
            auth_token: self.auth_token.clone(),
        };
        
        // Send discovery request
        self.send_message(&message).await?;
        
        // Wait for responses
        let start_time = SystemTime::now();
        let mut discovered = HashMap::new();
        
        // Subscribe to discovery events
        let mut event_receiver = self.event_sender.subscribe();
        
        while start_time.elapsed().unwrap_or(Duration::MAX) < discover_timeout {
            match timeout(Duration::from_millis(100), event_receiver.recv()).await {
                Ok(Ok(DiscoveryEvent::ServiceDiscovered { service_name, service_info })) => {
                    if service_names.is_empty() || service_names.contains(&service_name) {
                        discovered.insert(service_name, service_info);
                    }
                }
                Ok(Ok(DiscoveryEvent::ServiceUpdated { service_name, service_info })) => {
                    if service_names.is_empty() || service_names.contains(&service_name) {
                        discovered.insert(service_name, service_info);
                    }
                }
                _ => continue,
            }
            
            // Break if we found all requested services
            if !service_names.is_empty() && 
               service_names.iter().all(|name| discovered.contains_key(name)) {
                break;
            }
        }
        
        info!("Discovery completed, found {} services", discovered.len());
        Ok(discovered)
    }

    /// Send health ping
    pub async fn send_health_ping(
        &self,
        service_name: &str,
        status: &str,
        metrics: HashMap<String, String>,
    ) -> Result<(), NetworkError> {
        debug!("Sending health ping for service: {}", service_name);
        
        let message = DiscoveryMessage {
            message_type: DiscoveryMessageType::HealthPing,
            service_name: service_name.to_string(),
            timestamp: current_iso_timestamp(),
            payload: DiscoveryPayload::Health {
                status: status.to_string(),
                metrics,
            },
            auth_token: self.auth_token.clone(),
        };
        
        self.send_message(&message).await?;
        Ok(())
    }

    /// Announce service shutdown
    pub async fn announce_shutdown(&self, service_name: &str) -> Result<(), NetworkError> {
        info!("Announcing shutdown for service: {}", service_name);

        // Remove from local services
        {
            let mut local = self.local_services.write().await;
            local.remove(service_name);
        }

        let message = DiscoveryMessage {
            message_type: DiscoveryMessageType::ServiceShutdown,
            service_name: service_name.to_string(),
            timestamp: current_iso_timestamp(),
            payload: DiscoveryPayload::Empty,
            auth_token: self.auth_token.clone(),
        };

        self.send_message(&message).await?;
        Ok(())
    }

    /// Get currently discovered services from cache
    pub async fn get_cached_services(&self) -> HashMap<String, ServiceInfo> {
        let services_guard = self.discovered_services.read().await;
        services_guard.iter()
            .map(|(name, (info, _))| (name.clone(), info.clone()))
            .collect()
    }

    /// Subscribe to discovery events
    pub fn subscribe_events(&self) -> broadcast::Receiver<DiscoveryEvent> {
        self.event_sender.subscribe()
    }

    /// Send a message via multicast
    async fn send_message(&self, message: &DiscoveryMessage) -> Result<(), NetworkError> {
        self.send_message_to(message, self.multicast_addr).await
    }

    /// Send a message to a specific address (for unicast responses)
    async fn send_message_to(&self, message: &DiscoveryMessage, addr: SocketAddr) -> Result<(), NetworkError> {
        let socket_guard = self.socket.read().await;
        let socket = socket_guard.as_ref()
            .ok_or_else(|| NetworkError::DiscoveryFailed("Socket not initialized".to_string()))?;

        let message_data = serde_json::to_vec(message)?;
        socket.send_to(&message_data, addr)?;

        debug!("Sent {} message for service {} to {}",
               serde_json::to_string(&message.message_type).unwrap_or_default(),
               message.service_name,
               addr);

        Ok(())
    }

    /// Process received discovery message
    async fn _process_message(&self, message: DiscoveryMessage, sender_addr: SocketAddr) -> Result<(), NetworkError> {
        debug!("Processing {} message from {} for service {}", 
               serde_json::to_string(&message.message_type).unwrap_or_default(),
               sender_addr,
               message.service_name);

        // Verify authentication token if required
        if self.auth_token.is_some() && message.auth_token != self.auth_token {
            warn!("Authentication failed for message from {}", sender_addr);
            return Ok(()); // Ignore unauthorized messages
        }

        match message.message_type {
            DiscoveryMessageType::ServiceAnnouncement => {
                if let DiscoveryPayload::ServiceInfo(service_info) = message.payload {
                    self._cache_discovered_service(&message.service_name, service_info.clone()).await;
                    
                    let _ = self.event_sender.send(DiscoveryEvent::ServiceDiscovered {
                        service_name: message.service_name.clone(),
                        service_info,
                    });
                }
            }
            
            DiscoveryMessageType::DiscoveryRequest => {
                debug!("Received discovery request from {}", sender_addr);

                if let DiscoveryPayload::Request { service_names, .. } = &message.payload {
                    let local = self.local_services.read().await;

                    // Respond with matching local services
                    for (name, info) in local.iter() {
                        // If service_names is empty, respond with all; otherwise filter
                        if service_names.is_empty() || service_names.contains(name) {
                            let response = DiscoveryMessage {
                                message_type: DiscoveryMessageType::DiscoveryResponse,
                                service_name: name.clone(),
                                timestamp: current_iso_timestamp(),
                                payload: DiscoveryPayload::ServiceInfo(info.clone()),
                                auth_token: self.auth_token.clone(),
                            };

                            if let Err(e) = self.send_message_to(&response, sender_addr).await {
                                warn!("Failed to send discovery response to {}: {}", sender_addr, e);
                            }
                        }
                    }
                }
            }

            DiscoveryMessageType::DiscoveryResponse => {
                // Handle discovery response (same as announcement)
                if let DiscoveryPayload::ServiceInfo(service_info) = message.payload {
                    self._cache_discovered_service(&message.service_name, service_info.clone()).await;

                    let _ = self.event_sender.send(DiscoveryEvent::ServiceDiscovered {
                        service_name: message.service_name.clone(),
                        service_info,
                    });
                }
            }
            
            DiscoveryMessageType::HealthPing => {
                if let DiscoveryPayload::Health { status, metrics } = message.payload {
                    let _ = self.event_sender.send(DiscoveryEvent::HealthPing {
                        service_name: message.service_name,
                        status,
                        metrics,
                    });
                }
            }
            
            DiscoveryMessageType::ServiceShutdown => {
                self._remove_cached_service(&message.service_name).await;
                let _ = self.event_sender.send(DiscoveryEvent::ServiceLost {
                    service_name: message.service_name,
                });
            }
            
            _ => {
                debug!("Unhandled message type: {:?}", message.message_type);
            }
        }

        Ok(())
    }

    /// Cache a discovered service
    async fn _cache_discovered_service(&self, service_name: &str, service_info: ServiceInfo) {
        let mut services_guard = self.discovered_services.write().await;
        services_guard.insert(service_name.to_string(), (service_info, SystemTime::now()));
    }

    /// Remove cached service
    async fn _remove_cached_service(&self, service_name: &str) {
        let mut services_guard = self.discovered_services.write().await;
        services_guard.remove(service_name);
    }

    /// Start cache cleanup task to remove expired entries
    async fn start_cache_cleanup(&self) {
        let discovered_services = Arc::clone(&self.discovered_services);
        let event_sender = self.event_sender.clone();
        let cache_timeout = self.cache_timeout;
        
        tokio::spawn(async move {
            let mut interval = interval(Duration::from_secs(60)); // Clean every minute
            
            loop {
                interval.tick().await;
                
                let mut services_guard = discovered_services.write().await;
                let mut expired_services = Vec::new();
                
                services_guard.retain(|service_name, (_, last_seen)| {
                    if last_seen.elapsed().unwrap_or(Duration::MAX) > cache_timeout {
                        expired_services.push(service_name.clone());
                        false
                    } else {
                        true
                    }
                });
                
                drop(services_guard);
                
                // Send service lost events for expired services
                for service_name in expired_services {
                    let _ = event_sender.send(DiscoveryEvent::ServiceLost { service_name });
                }
            }
        });
    }
}

/// Get current ISO 8601 timestamp
fn current_iso_timestamp() -> String {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs()
        .to_string()
}